salt/utils/x509.py

```diff
@@ -0,0 +1,2002 @@
+                        padding.MGF1(cert.signature_hash_algorithm), padding.PSS.AUTO
+                    ),
+                    cert.signature_hash_algorithm,
+                )
+            else:
+                pubkey.verify(
+                    cert.signature,
+                    cert.tbs_certificate_bytes,
+                    padding.PKCS1v15(),
+                    cert.signature_hash_algorithm,
+                )
+            return True
+        except InvalidSignature:
+            return False
+    if key_type == KEY_TYPE.EC:
+        try:
+            pubkey.verify(
+                cert.signature,
+                cert.tbs_certificate_bytes,
+                ec.ECDSA(cert.signature_hash_algorithm),
+            )
+            return True
+        except InvalidSignature:
+            return False
+    if key_type in [KEY_TYPE.ED25519, KEY_TYPE.ED448]:
+        try:
+            pubkey.verify(cert.signature, cert.tbs_certificate_bytes)
+            return True
+        except InvalidSignature:
+            return False
+    raise SaltInvocationError(
+        "Invalid public key type, can only process rsa, ec, ed25519, ed448"
+    )
+
+
+def isfile(path):
+    """
+    A wrapper around os.path.isfile that ignores ValueError exceptions which
+    can be raised if the input to isfile is too long.
+    """
+    try:
+        return os.path.isfile(path)
+    except (TypeError, ValueError):
+        pass
+    return False
+
+
+def split_pems(pems):
+    """
+    Returns a list of PEM strings from a possibly concatenated one
+    """
+    pems = salt.utils.stringutils.to_bytes(pems)
+    splits = []
+    cur = []
+
+    for line in pems.splitlines(True):
+        if not line.strip():
+            continue
+        if line.startswith(PEM_BEGIN):
+            cur = [line]
+            continue
+        cur.append(line)
+        if line.startswith(PEM_END):
+            splits.append(b"".join(cur))
+            cur = []
+    return splits
+
+
+def load_file_or_bytes(fob):
+    """
+    Tries to load a reference and return its bytes.
+    Can be a file path on the local system, a string and bytes (hex/base64-encoded, raw)
+    """
+    if isfile(fob):
+        with salt.utils.files.fopen(fob, "rb") as f:
+            fob = f.read()
+    if isinstance(fob, str):
+        if PEM_BEGIN.decode() in fob:
+            fob = fob.encode()
+        else:
+            try:
+                fob = bytes.fromhex(fob)
+            except ValueError:
+                try:
+                    fob = base64.b64decode(fob)
+                except ValueError:
+                    pass
+    if not isinstance(fob, bytes):
+        raise SaltInvocationError(
+            "Could not load provided source. You need to pass an existing file, "
+            "(PEM|hex|base64)-encoded string or raw bytes."
+        )
+    return fob
+
+
+def _create_extension(name, val, subject_pubkey=None, ca_crt=None, ca_pub=None):
+    if name == "basicConstraints":
```

Comment:
This is a pretty big function. Any possibilities to break this up a bit to make development and testing easier?

Reply:
I refactored the `_create_extension` and `render_extension` catch-all functions into separate ones. Couldn't think of a natural way to group the render/build functions of each one together, so I just broke out the pieces.